patches/tModLoader/Terraria.World.Generation/WorldGenerator.cs.patch

```diff
@@ -13,4 +13,28 @@
  
  		public WorldGenerator(int seed)
  		{
+@@ -44,7 +_,22 @@
+ 				Main.rand = new UnifiedRandom(this._seed);
+ 				stopwatch.Start();
+ 				progress.Start(current2.Weight);
+-				current2.Apply(progress);
++				try
++				{
++					current2.Apply(progress);
++				}
++				catch(Exception e)
++				{
++					string message = string.Join
++					(
++						"\n",
++						Localization.Language.GetTextValue("tModLoader.WorldGenError"),
++						current2.Name,
++						e.Message
++					);
++					Main.MenuUI.SetState(new ModLoader.UI.UIWorldLoadCrash(message));
```

Comment:
Is this threadsafe? I do much prefer constructing UI's each time they're used, but it's not the Terraria paradigm.

Reply:
Shouldn't it be setting Main.menuMode to a custom ID? That is how we do this for all menus. Check Interface.cs
( I do want to rewrite this, still )